### ISSUE TITLE
make pageMargins dynamic

### ISSUE THREAD

**Issue by Pczek** [bpampuch/pdfmake#368]
Books are binded and therefore have different page margins for odd and even pages, so that the text on the page appears to be in the middle and not squeezed into the books fold. I searched the documentation/issues and tested it, the pageMargins seem to be static.

**Comment by nthnmldr** [comment]
Yes, would be very helpful to be dynamic.  I'd like to generate a different report footer just for the last page.  It would nice to have pageMargins accept a function just like header/footer that has two parameters of currentPage and pageCount.

**Comment by Nor145** [comment]
+1

**Comment by oscardz10** [comment]
+1

**Comment by max-yu** [comment]
+1

**Comment by difnicolas** [comment]
+1

**Comment by GuidoKoepke** [comment]
+1

**Comment by kngchn** [comment]
+1

**Comment by cbernecker** [comment]
+1

**Comment by uNki23** [comment]
+1

**Comment by tkbrady** [comment]
+1

**Comment by icke792** [comment]
i solve my problem with different header heights changing follow functions in documentContext.js.

```
function DocumentContext(pageSize, pageMargins) {
	this.pages = [];

	this.pageMargins = pageMargins;

      this.pageMarginsFkt = null;

        if (typeof this.pageMargins === 'function'){
             this.pageMarginsFkt = this.pageMargins; 
        }

        if (this.pageMarginsFkt){
             this.pageMargins = this.pageMarginsFkt(1);
        }
	this.x =this. pageMargins.left;
	this.availableWidth = pageSize.width - this.pageMargins.left - this.pageMargins.right;
	this.availableHeight = 0;
	this.page = -1;

	this.snapshots = [];

	this.endingCell = null;

	this.tracker = new TraversalTracker();

	this.addPage(pageSize);

	this.hasBackground = false;
}
```

```
`DocumentContext.prototype.initializePage = function () {
	this.y = this.getCurrentPage().pageMargins.top;
	this.availableHeight = this.getCurrentPage().pageSize.height - this.getCurrentPage().pageMargins.top                    - this.getCurrentPage().pageMargins.bottom;
	this.pageSnapshot().availableWidth = this.getCurrentPage().pageSize.width - this.getCurrentPage().pageMargins.left - this.getCurrentPage().pageMargins.right;
};`
```

```
`DocumentContext.prototype.addPage = function (pageSize) {
	var page = {items: [], pageSize: pageSize};
	this.pages.push(page);
	this.page = this.pages.length - 1;

        if (typeof this.pageMargins === 'function'){
             this.pageMarginsFkt = this.pageMargins; 
        }

        if (this.pageMarginsFkt){
             this.pageMargins = this.pagerMarginsFkt(this.pages.length);
        }

        page.pageMargins = Object.assign( {} , this.pageMargins );

	this.initializePage();

	this.tracker.emit('pageAdded');

	return page;
};`
```

```
`DocumentContext.prototype.getCurrentPosition = function () {
	var pageSize = this.getCurrentPage().pageSize;
	var innerHeight = pageSize.height - this.getCurrentPage().pageMargins.top - this.getCurrentPage().pageMargins.bottom;
	var innerWidth = pageSize.width - this.getCurrentPage().pageMargins.left - this.getCurrentPage().pageMargins.right;

	return {
               pageMargins: this.getCurrentPage.pageMargins,
		pageNumber: this.page + 1,
		pageOrientation: pageSize.orientation,
		pageInnerHeight: innerHeight,
		pageInnerWidth: innerWidth,
		left: this.x,
		top: this.y,
		verticalRatio: ((this.y - this.pageMargins.top) / innerHeight),
		horizontalRatio: ((this.x - this.pageMargins.left) / innerWidth)
	};
};`
```

so i can set a function in docDefinition for pageMargins like:

```
`var docDefinition = {
           ......,
           pageMargins : function(cp) {
                  if ( cp === 1 ){
                         return {left:30,top:150, right:30,bottom:60};
                  } else {
                         return {left:30,top:50, right:30,bottom:60};
                  }
           },
          .......
}`
```
I have not found any negative side effects for me and I hope to have listed all my changes (I got them from a printout). Maybe that solves the problem of one or the other. Happy Easter :-)

sorry for my english, isn't my favorit language.

**Comment by renarsvilnis** [comment]
@icke792 Can you create a pull-request for it?

**Comment by hobbsi** [comment]
This was really useful for me. Please add it to the project. There seems to be a small fix to make it work with dynamic horizontal margins:

DocumentContext.prototype.initializePage = function () {
	this.y = this.getCurrentPage().pageMargins.top;
	this.x = this.getCurrentPage().pageMargins.left;
	this.availableHeight = this.getCurrentPage().pageSize.height - this.getCurrentPage().pageMargins.top                    - this.getCurrentPage().pageMargins.bottom;
	this.availableWidth = this.getCurrentPage().pageSize.width - this.getCurrentPage().pageMargins.left                    - this.getCurrentPage().pageMargins.right;
};

**Comment by kunalwatkar** [comment]
+1

**Comment by zenway-js** [comment]
+1

**Comment by Corassa** [comment]
+1

**Comment by naknode** [comment]
> This was really useful for me. Please add it to the project. There seems to be a small fix to make it work with dynamic horizontal margins:
> 
> ```
> DocumentContext.prototype.initializePage = function () {
> 	this.y = this.getCurrentPage().pageMargins.top;
> 	this.x = this.getCurrentPage().pageMargins.left;
> 	this.availableHeight = this.getCurrentPage().pageSize.height - this.getCurrentPage().pageMargins.top                    - this.getCurrentPage().pageMargins.bottom;
> 	this.availableWidth = this.getCurrentPage().pageSize.width - this.getCurrentPage().pageMargins.left                    - this.getCurrentPage().pageMargins.right;
> };
> ```

@hobbsi Did you just edit the source, build it and roll your own version?

**Comment by hobbsi** [comment]
> @hobbsi Did you just edit the source, build it and roll your own version?

I used it while doing some experimentation on a personal project. And yes, that's what did. I believe you could also fork the project and do your modifications... I didn't.

**Comment by hielfx** [comment]
> This was really useful for me. Please add it to the project. There seems to be a small fix to make it work with dynamic horizontal margins:
> 
> ```
> DocumentContext.prototype.initializePage = function () {
> 	this.y = this.getCurrentPage().pageMargins.top;
> 	this.x = this.getCurrentPage().pageMargins.left;
> 	this.availableHeight = this.getCurrentPage().pageSize.height - this.getCurrentPage().pageMargins.top                    - this.getCurrentPage().pageMargins.bottom;
> 	this.availableWidth = this.getCurrentPage().pageSize.width - this.getCurrentPage().pageMargins.left                    - this.getCurrentPage().pageMargins.right;
> };
> ```

How would you apply this, once implemented, in the document definition?

**Comment by robstoll** [comment]
@liborm85 any chance the approach of @icke792  with the fix of @hobbsi gets merged if I would create a PR?

Also, I have the need for a dynamic margin based on a dynamic header (see #2238) which is even a step further than what we have here. Any chance a solution taking dynamic headers into account gets merged in the end?

**Comment by gine** [comment]
A work around can be setting the content margins to negative value  (only top margin) and passing an empty string to the header function.

For example:

pageMargins : [30, 140, 30, 30]

Then on the page where you don't need header, after passed the empty string, you can sett:

content.margins: [0, -110, 0, 0 ]

**Comment by qasimmehdi** [comment]
> A work around can be setting the content margins to negative value (only top margin) and passing an empty string to the header function.
> 
> For example:
> 
> pageMargins : [30, 140, 30, 30]
> 
> Then on the page where you don't need header, after passed the empty string, you can sett:
> 
> content.margins: [0, -110, 0, 0 ]

Will it work for footers?

**Comment by qasimmehdi** [comment]
If anyone wants to create bigger footer on a specific page you can just put your items in the footer with page condition and give it `margin: [0,-100,0,0]` decrease the negative value to fit your content.
Note: this is just a workaround and if you have more text in your page body using negative margin might make footer and body overlap.

**Comment by big-scorpio** [comment]
> Books are binded and therefore have different page margins for odd and even pages, so that the text on the page appears to be in the middle and not squeezed into the books fold. I searched the documentation/issues and tested it, the pageMargins seem to be static.



> Books are binded and therefore have different page margins for odd and even pages, so that the text on the page appears to be in the middle and not squeezed into the books fold. I searched the documentation/issues and tested it, the pageMargins seem to be static.

Did you solve the issue?

**Comment by big-scorpio** [comment]
I solved this issue, contact me if someone need help
zhengwei@byartifact.com

**Comment by albertoLGDev** [comment]
> 

Has anyone found a solution for this?

**Comment by carltin0315** [comment]
Can I determine It's the last page of the pdf and assign different page margin for it?

**Comment by GitNomster** [comment]
+1

**Comment by Kritsana135** [comment]
+ 1

**Comment by Cmenesess** [comment]
+1

**Comment by scottcorgan** [comment]
+2

**Comment by Matoseb** [comment]
+1

**Comment by justlester** [comment]
+1
